writers/postgres/messages.go

```diff
@@ -97,57 +104,94 @@ func (pr postgresRepo) saveSenml(messages interface{}) error {
 	return err
 }
 
-func (pr postgresRepo) saveJSON(messages interface{}) error {
-	msg, ok := messages.(mfjson.Message)
-	if !ok {
-		return errSaveMessage
+func (pr postgresRepo) saveJSON(msgs mfjson.Messages) error {
+	if err := pr.insertJSON(msgs); err != nil {
+		if err == errNoTable {
+			if err := pr.createTable(msgs.Format); err != nil {
+				return err
+			}
+			return pr.insertJSON(msgs)
+		}
+		return err
 	}
-	q := `INSERT INTO json (id, channel, subtopic, publisher, protocol, payload)
-          VALUES (:id, :channel, :subtopic, :publisher, :protocol, :payload);`
+	return nil
+}
 
-	dbmsg, err := toJSONMessage(msg)
+func (pr postgresRepo) insertJSON(msgs mfjson.Messages) error {
+	tx, err := pr.db.BeginTxx(context.Background(), nil)
 	if err != nil {
 		return errors.Wrap(errSaveMessage, err)
 	}
-
-	if _, err := pr.db.NamedExec(q, dbmsg); err != nil {
-		pqErr, ok := err.(*pq.Error)
-		if ok {
-			switch pqErr.Code.Name() {
-			case errInvalid:
-				return errors.Wrap(errSaveMessage, ErrInvalidMessage)
+	defer func() {
+		if err != nil {
+			if txErr := tx.Rollback(); txErr != nil {
+				err = errors.Wrap(err, errors.Wrap(errTransRollback, txErr))
 			}
+			return
 		}
 
-		return errors.Wrap(errSaveMessage, err)
+		if err = tx.Commit(); err != nil {
+			err = errors.Wrap(errSaveMessage, err)
+		}
+		return
+	}()
+
+	q := `INSERT INTO %s (id, channel, created, subtopic, publisher, protocol, payload)
```

Comment:
can this be in one line

Reply:
Unlike [in this case](https://github.com/mainflux/mainflux/pull/1279#discussion_r526319716), we're using `NamedExec` here, so the rest of the query is much larger and can't fit in a single line.

Reply:
actually I meant together with this `q = fmt.Sprintf(q, msgs.Format)`

Reply:
I find it a bit easier to read this way:
```go
q := `INSERT INTO %s (id, channel, created, subtopic, publisher, protocol, payload)
      VALUES (:id, :channel, :created, :subtopic, :publisher, :protocol, :payload);`
q = fmt.Sprintf(q, msgs.Format)
```
than this:
```go
q = fmt.Sprintf(`INSERT INTO %s (id, channel, created, subtopic, publisher, protocol, payload) 
                 VALUES (:id, :channel, :created, :subtopic, :publisher, :protocol, :payload);`, msgs.Format)
```
If you insist, I'll use the second option.